core/src/test/java/io/parsingdata/metal/AutoEqualityTest.java

```diff
@@ -0,0 +1,283 @@
+import io.parsingdata.metal.data.ConstantSource;
+import io.parsingdata.metal.data.DataExpressionSource;
+import io.parsingdata.metal.data.Environment;
+import io.parsingdata.metal.data.ParseGraph;
+import io.parsingdata.metal.data.ParseItem;
+import io.parsingdata.metal.data.ParseReference;
+import io.parsingdata.metal.data.ParseValue;
+import io.parsingdata.metal.data.Slice;
+import io.parsingdata.metal.data.Source;
+import io.parsingdata.metal.encoding.Encoding;
+import io.parsingdata.metal.expression.Expression;
+import io.parsingdata.metal.expression.comparison.Eq;
+import io.parsingdata.metal.expression.comparison.EqNum;
+import io.parsingdata.metal.expression.comparison.EqStr;
+import io.parsingdata.metal.expression.comparison.GtNum;
+import io.parsingdata.metal.expression.comparison.LtNum;
+import io.parsingdata.metal.expression.value.Cat;
+import io.parsingdata.metal.expression.value.Const;
+import io.parsingdata.metal.expression.value.ConstantFactory;
+import io.parsingdata.metal.expression.value.Elvis;
+import io.parsingdata.metal.expression.value.FoldLeft;
+import io.parsingdata.metal.expression.value.FoldRight;
+import io.parsingdata.metal.expression.value.Reverse;
+import io.parsingdata.metal.expression.value.Value;
+import io.parsingdata.metal.expression.value.ValueExpression;
+import io.parsingdata.metal.expression.value.arithmetic.Add;
+import io.parsingdata.metal.expression.value.arithmetic.Div;
+import io.parsingdata.metal.expression.value.arithmetic.Mod;
+import io.parsingdata.metal.expression.value.arithmetic.Mul;
+import io.parsingdata.metal.expression.value.arithmetic.Neg;
+import io.parsingdata.metal.expression.value.bitwise.And;
+import io.parsingdata.metal.expression.value.bitwise.Not;
+import io.parsingdata.metal.expression.value.bitwise.Or;
+import io.parsingdata.metal.expression.value.bitwise.ShiftLeft;
+import io.parsingdata.metal.expression.value.bitwise.ShiftRight;
+import io.parsingdata.metal.expression.value.reference.Count;
+import io.parsingdata.metal.expression.value.reference.First;
+import io.parsingdata.metal.expression.value.reference.Last;
+import io.parsingdata.metal.expression.value.reference.Len;
+import io.parsingdata.metal.expression.value.reference.NameRef;
+import io.parsingdata.metal.expression.value.reference.Nth;
+import io.parsingdata.metal.expression.value.reference.Offset;
+import io.parsingdata.metal.token.Cho;
+import io.parsingdata.metal.token.Def;
+import io.parsingdata.metal.token.Nod;
+import io.parsingdata.metal.token.Opt;
+import io.parsingdata.metal.token.Pre;
+import io.parsingdata.metal.token.Rep;
+import io.parsingdata.metal.token.RepN;
+import io.parsingdata.metal.token.Seq;
+import io.parsingdata.metal.token.Sub;
+import io.parsingdata.metal.token.Tie;
+import io.parsingdata.metal.token.Token;
+import io.parsingdata.metal.token.TokenRef;
+import io.parsingdata.metal.token.While;
+import io.parsingdata.metal.util.InMemoryByteStream;
```

Comment:
With this types of classes I would suggest not to use imports, but the lines are already long as it is...